src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/FunctionProbes/FunctionProbesManager.cs

```diff
@@ -25,50 +28,176 @@ internal sealed class FunctionProbesManager : IFunctionProbesManager
             uint count,
             [MarshalAs(UnmanagedType.LPArray)] uint[] boxingTokens,
             [MarshalAs(UnmanagedType.LPArray)] uint[] boxingTokenCounts);
-       
-        private readonly object _requestLocker = new();
+
+        private delegate void FunctionProbeRegistrationCallback(int hresult);
+        private delegate void FunctionProbeInstallationCallback(int hresult);
+        private delegate void FunctionProbeUninstallationCallback(int hresult);
+        private delegate void FunctionProbeFaultCallback(ulong uniquifier);
+
+        [DllImport(ProfilerIdentifiers.MutatingProfiler.LibraryRootFileName, CallingConvention = CallingConvention.StdCall, PreserveSig = false)]
+        private static extern void RegisterFunctionProbeCallbacks(
+            FunctionProbeRegistrationCallback onRegistration,
+            FunctionProbeInstallationCallback onInstallation,
+            FunctionProbeUninstallationCallback onUninstallation,
+            FunctionProbeFaultCallback onFault);
+
+        [DllImport(ProfilerIdentifiers.MutatingProfiler.LibraryRootFileName, CallingConvention = CallingConvention.StdCall, PreserveSig = false)]
+        private static extern void UnregisterFunctionProbeCallbacks();
+
+        private long _probeState;
+        private const long ProbeStateUninitialized = default(long);
+        private const long ProbeStateUninstalled = 1;
+        private const long ProbeStateUninstalling = 2;
+        private const long ProbeStateInstalling = 3;
+        private const long ProbeStateInstalled = 4;
+        private const long ProbeStateUnrecoverable = 5;
+
+        private readonly TaskCompletionSource _probeRegistrationTaskSource = new(TaskCreationOptions.RunContinuationsAsynchronously);
+        private TaskCompletionSource? _installationTaskSource;
+        private TaskCompletionSource? _uninstallationTaskSource;
+
         private long _disposedState;
-        private bool _isCapturing;
+
+        public event EventHandler<InstrumentedMethod>? OnProbeFault;
 
         public FunctionProbesManager(IFunctionProbes probes)
         {
             ProfilerResolver.InitializeResolver<FunctionProbesManager>();
 
+            RegisterFunctionProbeCallbacks(OnRegistration, OnInstallation, OnUninstallation, OnFault);
```

Comment:
Persisted pointers to functions across managed to native boundaries usually require pinning. OnRegistration will be pinned by the marshaller for the duration of the RegisterFunctionProbeCallbacks call, but after that function ends, the address may get moved around in memory during garbage collection.